### PR TITLE
Samtools cram bam conversion wrappers

### PR DESCRIPTION
**Pull request by yhoogstrate**
This is the second attempt to get BAM `<->` CRAM conversion tools into galaxy. This wrapper uses samtools 1.2 instead of cramtools.

### CODE REVIEW

File: tools/samtools_bam_to_cram/samtools_bam_to_cram.xml [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+        <exit_code range="1:" />
+        <exit_code range=":-1" />
+    </stdio>
+
+    <version_command><![CDATA[ samtools 2>&1 | grep Version ]]></version_command>
+
+    <command>
+    <![CDATA[
+        ## If no bam index exists, e.g. using planemo test, create one:
+        #if $parameter_regions.target_region != "entire_input_file"
```
Comment (bgruening) [inline tools/samtools_bam_to_cram/samtools_bam_to_cram.xml]: This should not be a problem. During upload in Galaxy an index will be created.

Comment (yhoogstrate) [inline tools/samtools_bam_to_cram/samtools_bam_to_cram.xml]: Thanks @bgruening, I knew it was supposed to work but I didn't. I have tracked the problem down. The internal code of galaxy uses `samtools index dataset.dat dataset.dat.bai`. Unfortunately samtools 1.2 gives a coredump if the `[out.index]` is defined. I will update the PR but the tests that use defined regions might fail. I will see if I can fix or report this issue upstream.

File: tools/samtools_bam_to_cram/samtools_bam_to_cram.xml [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+        </conditional>
+        <conditional name="parameter_regions">
+            <param name="target_region" type="select" label="Choose conversion within specific genomic region(s)">
+                <option value="entire_input_file">Entire BAM alignment file</option>
+                <option value="region">Specific region</option>
+                <option value="regions_bed_file">List of specific regions (BED file)</option>
+            </param>
+            <when value="entire_input_file" />
+            <when value="region">
+                <param name="region_string" type="text" label="Samtools: region in which pileup is generated" help="e.g. chrX or chr:start-end" />
```
Comment (bgruening) [inline tools/samtools_bam_to_cram/samtools_bam_to_cram.xml]: regarding your santizing question: I think it would make much sense to have a global genomic_region validator for chr:start-end.

File: tools/samtools_cram_to_bam/samtools_cram_to_bam.xml [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+        <exit_code range="1:" />
+        <exit_code range=":-1" />
+    </stdio>
+
+    <version_command><![CDATA[ samtools 2>&1 | grep Version ]]></version_command>
+
+    <command>
+    <![CDATA[
+        ## If no cram index exists, e.g. using planemo test, create one:
+        #if $parameter_regions.target_region != "entire_input_file"
```
Comment (bgruening) [inline tools/samtools_cram_to_bam/samtools_cram_to_bam.xml]: This one should be removed as well. If planemo crashes during index build, Galaxy will also crash. And we should fix this upstream and exclude the option unless this is fixed. Otherwise we will not have any metadata, correct?

File: tools/samtools_cram_to_bam/samtools_cram_to_bam.xml [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+        samtools view
+            #if $parameter_regions.target_region == "regions_bed_file"
+                -L parameter_regions.regions_bed_file
+            #end if
+            
+            -b
+            -T ${reference_source.input_reference}
+            -o $output_alignment
+            $input_alignment
+
```
Comment (bgruening) [inline tools/samtools_cram_to_bam/samtools_cram_to_bam.xml]: Is the @ parameter supported here as well to speed up the conversion process with multiple threads?

**Comment by bgruening** [comment]
I like this one very much! Thanks @yhoogstrate for considering samtools!

File: tools/samtools_cram_to_bam/samtools_cram_to_bam.xml (hunk 3 above)
Comment (yhoogstrate) [inline tools/samtools_cram_to_bam/samtools_cram_to_bam.xml]: Samtools index in 1.2 will only crash if a 4th argument (path to the index) is specified. E.g. `samtools index "${input_alignment}" "${input_alignment}.crai"`. It works as given above, so this line should be preserved since cram indices (*.crai) are not part of the current cram datatype in galaxy.

The ideal situation would be to include .crai cram indexes within the galaxy datatype definition, which would make this if-statement unnecessary. I have thought of implementing this, but adding a cram index via a samtools background process becomes a big mess because of the different versions of samtools. Galaxy should come with 1 fixed version of samtools to make this workable. Currently the galaxy datatype definition of bam indices already contains different lines of code for specific versions of samtools. Additional lines would also be required for cram indices (to support samtools or 1.2), etc. So my suggestion would be to include the lines above until crai indices are part of the galaxy datatypes.

File: tools/samtools_cram_to_bam/samtools_cram_to_bam.xml (hunk 4 above)
Comment (yhoogstrate) [inline tools/samtools_cram_to_bam/samtools_cram_to_bam.xml]: Good one

File: tools/samtools_bam_to_cram/samtools_bam_to_cram.xml (hunk 2 above)
Comment (yhoogstrate) [inline tools/samtools_bam_to_cram/samtools_bam_to_cram.xml]: Good one, will take a look at this

**Comment by nsoranzo** [comment]
Would it make more sense to add these to https://github.com/galaxyproject/tools-devteam/tree/master/tool_collections/samtools since the other Samtools wrappers are going to be maintained there?

**Comment by yhoogstrate** [comment]
I am not sure whether the cram converters are considered as new tools?

File: tools/samtools_bam_to_cram/samtools_bam_to_cram.xml [hunk 5]
```diff
@@ -0,0 +1,137 @@
+<tool id="samtools_bam_to_cram" name="samtools BAM to CRAM" version="1.2.a">
```
Comment (bgruening) [inline tools/samtools_bam_to_cram/samtools_bam_to_cram.xml]: I would prefer only numbers in the version string.

**Comment by bgruening** [comment]
@nsoranzo good question. We have the policy to not add new tools to devteam. On the otherhand it would be nice to have all samtools wrappers in one place to make bulk updates possible.

I'm +/- 0 here. @erasche?
No matter what we decide we should add these tools to the suite under devteam, imho.

**Comment by nsoranzo** [comment]
Yes, we have this policy, but this not a standalone tool, it's part of a tool collection.

**Comment by yhoogstrate** [comment]
WIP is pending on an update to the CRAM datatype

**Comment by bgruening** [comment]
@yhoogstrate I guess we should merge this in January with the next release?

**Comment by erasche** [comment]
yeah, might as well add to devteam. It's really, really hard to follow that policy even though I think it makes sense since most of "devteam" aren't really producing the bulk of the tools these days.

**Comment by bgruening** [comment]
@yhoogstrate ok I think it would be nice to move this to devteam and merge it for the next release in January?

**Comment by yhoogstrate** [comment]
@bgruening Understood! I will do that when the datatype PR gets in. Thanks for the reviews guys.
